lib/Dialect/AIE/Transforms/AIEAssignLockIDs.cpp

```diff
@@ -40,50 +39,63 @@ struct AIEAssignLockIDsPass : AIEAssignLockIDsBase<AIEAssignLockIDsPass> {
     DeviceOp device = getOperation();
     OpBuilder rewriter = OpBuilder::atBlockEnd(device.getBody());
 
-    std::map<Operation *, std::pair<int, std::set<int>>> tileToLastID;
-
-    // The first pass scans for and stores the existing lockIDs. This data is
-    // stored in a map with the lock’s tile operation as the key, while the
-    // value to the key is a pair with the current potential lockID and a set
-    // that stores the currently assigned lockIDs.
-    for (auto lock : device.getOps<LockOp>()) {
-      if (lock.getLockID().has_value()) {
-        Operation *lock_tile = lock.getTile().getDefiningOp();
-        tileToLastID[lock_tile].first = 0;
-        tileToLastID[lock_tile].second.insert(lock.getLockIDValue());
+    // Map from tile to all its locks.
+    DenseMap<TileOp, SmallVector<LockOp>> tileToLockOps;
+    for (auto lockOp : device.getOps<LockOp>()) {
+      auto tileOp = lockOp.getTileOp();
+      auto iter = tileToLockOps.find(tileOp);
+      if (iter == tileToLockOps.end()) {
+        tileToLockOps.insert({tileOp, {lockOp}});
+      } else {
+        iter->second.push_back(lockOp);
```

Comment:
Could be DenseSet instead of SmallVector, then you wouldn't need to do the collision detection below?

Reply:
Right... but why have a vector *and* a set... why not just have a set?

Reply:
I've updated the design to avoid this initial vector. It's a bit more complex, but probably more efficient (only possible downside it that previously there was only 1 DenseSet, in the inner loop, now there is 1 per tile).